File: fact.nim
#  @@@ get_version: nim --version 2>&1 | awk 'NR == 1 {print $4}'
#  @@@ before: nim c fact.nim
#  @@@ instead: ./fact
#  @@@ after: rm -rf fact nimcache
#  @@@ is_fast: 1

# with:
#   nim c -d:release fact.nim
# programm fast with any "times"

const times = 10_000_000
const fact_16 = 20_922_789_888_000

proc fact(n: int): int =
  if n == 1:
      1
  else:
      n * fact(n - 1)

proc main() =
    var ok = true
    for i in 0..times:
        ok = ok and fact(16) == fact_16
    echo("nim finish ", times, " - ", if ok: "ok" else: "fail");

main()

File: fact.pas
(*
    @@@ get_version: fpc -h | awk 'NR == 1 {print $5}'
    @@@ before: fpc -O4 fact.pas
    @@@ instead: ./fact
    @@@ after: rm fact fact.o
    @@@ is_fast: 1
*)

program fact;

function fact(n: Int64): Int64;
begin
  if n = 0
  then
    fact := 1
  else
    fact := n * fact(n - 1)
end;

const
  times: Int64 = 100000000;
  fact_16: Int64 = 20922789888000;

var
  i: Int64;
  ok: boolean;
  res: string;

begin
  ok := True;
  for i := 1 to times do
  begin
    ok := ok and (fact_16 = fact(16));
  end;

  if ok then res := 'ok' else res := 'fail';
  WriteLn('Pascal finish ', times, ' - ', res);
end.
